File: go/handlers/storage.go
package handlers

import (
    "chat-app/models"
    "fmt"
    "strings"
    "sync"

    "github.com/google/uuid"
)

type Storage struct {
    users    map[uuid.UUID]models.User
    messages []models.Message
    mu       sync.RWMutex
}

func NewStorage() *Storage {
    return &Storage{
        users:    make(map[uuid.UUID]models.User),
        messages: make([]models.Message, 0),
    }
}

func (s *Storage) AddUser(user models.User) error {
    s.mu.Lock()
    defer s.mu.Unlock()

    for _, existingUser := range s.users {
        if existingUser.Username == user.Username {
            return fmt.Errorf("username already exists")
        }
    }

    s.users[user.ID] = user
    return nil
}

func (s *Storage) GetUser(id uuid.UUID) (models.User, bool) {
    s.mu.RLock()
    defer s.mu.RUnlock()

    user, exists := s.users[id]
    return user, exists
}

func (s *Storage) GetUserByUsername(username string) (models.User, bool) {
    s.mu.RLock()
    defer s.mu.RUnlock()

    for _, user := range s.users {
        if user.Username == username {
            return user, true
        }
    }
    return models.User{}, false
}

func (s *Storage) AddMessage(message models.Message) {
    s.mu.Lock()
    defer s.mu.Unlock()

    s.messages = append(s.messages, message)
}

func (s *Storage) GetMessagesForUser(userID uuid.UUID) []models.Message {
    s.mu.RLock()
    defer s.mu.RUnlock()

    var userMessages []models.Message
    for _, msg := range s.messages {
        if msg.SenderID == userID || msg.ReceiverID == userID {
            userMessages = append(userMessages, msg)
        }
    }
    return userMessages
}

func (s *Storage) SearchMessages(keyword string) []models.Message {
    s.mu.RLock()
    defer s.mu.RUnlock()

    var matchingMessages []models.Message
    keyword = strings.ToLower(keyword)
    for _, msg := range s.messages {
        if strings.Contains(strings.ToLower(msg.Content), keyword) {
            matchingMessages = append(matchingMessages, msg)
        }
    }
    return matchingMessages
}

func (s *Storage) ListUsers() []models.User {
    s.mu.RLock()
    defer s.mu.RUnlock()

    users := make([]models.User, 0, len(s.users))
    for _, user := range s.users {
        users = append(users, user)
    }
    return users
} 

File: rust/src/storage.rs
use crate::models::{Message, User};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use uuid::Uuid;

#[derive(Default)]
pub struct Storage {
    users: HashMap<Uuid, User>,
    messages: Vec<Message>,
}

impl Storage {
    pub fn new() -> Arc<Mutex<Self>> {
        Arc::new(Mutex::new(Self {
            users: HashMap::new(),
            messages: Vec::new(),
        }))
    }

    pub fn add_user(&mut self, user: User) -> Result<(), String> {
        if self.users.values().any(|u| u.username == user.username) {
            return Err("Username already exists".to_string());
        }
        self.users.insert(user.id, user);
        Ok(())
    }

    pub fn get_user(&self, id: &Uuid) -> Option<User> {
        self.users.get(id).cloned()
    }

    pub fn get_user_by_username(&self, username: &str) -> Option<User> {
        self.users
            .values()
            .find(|u| u.username == username)
            .cloned()
    }

    pub fn add_message(&mut self, message: Message) {
        self.messages.push(message);
    }

    pub fn get_messages_for_user(&self, user_id: &Uuid) -> Vec<Message> {
        self.messages
            .iter()
            .filter(|m| m.sender_id == *user_id || m.receiver_id == *user_id)
            .cloned()
            .collect()
    }

    pub fn search_messages(&self, keyword: &str, user_id: &Uuid) -> Vec<Message> {
        self.messages
            .iter()
            .filter(|m| m.sender_id == *user_id || m.receiver_id == *user_id)
            .filter(|m| m.content.to_lowercase().contains(&keyword.to_lowercase()))
            .cloned()
            .collect()
    }

    pub fn list_users(&self) -> Vec<User> {
        self.users.values().cloned().collect()
    }
} 